### ISSUE TITLE
Nomad dispatch jobs are restarted when they should not be

### ISSUE THREAD

**Issue by kurtwheeler** [hashicorp/nomad#4700]
### Nomad version
Nomad v0.8.3 (c85483da3471f4bd3a7c3de112e95f551071769f)

### Operating system and Environment details

We are running this within AWS using an ubuntu AMI:
```
ubuntu-16.04-docker-17.12.0-ce-201802201241 (ami-dd08efa0)
```

### Issue

I am running dispatch jobs and I have tried to configure them to not be restarted, because I have my own system for restarting jobs. Each job should only ever be started once and I have checks in my code to make sure of this. These checks are being triggered and it looks like it's because Nomad is restarting allocations, even though when the first allocation fails its last event is: `Not Restarting  Policy allows no restarts`.

```
ubuntu@ip-10-0-145-60:~$ nomad status SALMON_19_8192/dispatch-1537385599-1bf6edab
ID            = SALMON_19_8192/dispatch-1537385599-1bf6edab
Name          = SALMON_19_8192/dispatch-1537385599-1bf6edab
Submit Date   = 2018-09-19T19:33:19Z
Type          = batch
Priority      = 50
Datacenters   = dc1
Status        = dead
Periodic      = false
Parameterized = false

Summary
Task Group  Queued  Starting  Running  Failed  Complete  Lost
jobs        0       0         0        1       1         0

Allocations
ID        Node ID   Task Group  Version  Desired  Status    Created     Modified
665515bc  a16c93b7  jobs        0        run      complete  17m34s ago  13m56s ago
698d8d80  a16c93b7  jobs        0        run      failed    23m56s ago  16m8s ago
ubuntu@ip-10-0-145-60:~$ nomad status 698d8d80
ID                   = 698d8d80
Eval ID              = 5d4cce7f
Name                 = SALMON_19_8192/dispatch-1537385599-1bf6edab.jobs[0]
Node ID              = a16c93b7
Job ID               = SALMON_19_8192/dispatch-1537385599-1bf6edab
Job Version          = 0
Client Status        = failed
Client Description   = <none>
Desired Status       = run
Desired Description  = <none>
Created              = 24m2s ago
Modified             = 16m14s ago
Replacement Alloc ID = 665515bc

Task "salmon" is "dead"
Task Resources
CPU       Memory   Disk    IOPS  Addresses
1024 MHz  8.0 GiB  10 MiB  0     

Task Events:
Started At     = 2018-09-19T19:36:26Z
Finished At    = 2018-09-19T19:39:37Z
Total Restarts = 0
Last Restart   = N/A

Recent Events:
Time                  Type            Description
2018-09-19T19:39:37Z  Not Restarting  Policy allows no restarts
2018-09-19T19:39:37Z  Terminated      Exit Code: 0, Exit Message: "OOM Killed"
2018-09-19T19:36:26Z  Started         Task started by client
2018-09-19T19:33:30Z  Driver          Downloading image wkurt/dr_salmon:OOM
2018-09-19T19:33:24Z  Task Setup      Building Task Directory
2018-09-19T19:33:24Z  Received        Task received by client
ubuntu@ip-10-0-145-60:~$ nomad status 665515bc
ID                  = 665515bc
Eval ID             = 9aa2e115
Name                = SALMON_19_8192/dispatch-1537385599-1bf6edab.jobs[0]
Node ID             = a16c93b7
Job ID              = SALMON_19_8192/dispatch-1537385599-1bf6edab
Job Version         = 0
Client Status       = complete
Client Description  = <none>
Desired Status      = run
Desired Description = <none>
Created             = 17m44s ago
Modified            = 14m6s ago
Reschedule Attempts = 1/1

Task "salmon" is "dead"
Task Resources
CPU       Memory   Disk    IOPS  Addresses
1024 MHz  8.0 GiB  10 MiB  0     

Task Events:
Started At     = 2018-09-19T19:42:42Z
Finished At    = 2018-09-19T19:43:17Z
Total Restarts = 0
Last Restart   = N/A

Recent Events:
Time                  Type        Description
2018-09-19T19:43:17Z  Terminated  Exit Code: 0
2018-09-19T19:42:42Z  Started     Task started by client
2018-09-19T19:41:19Z  Task Setup  Building Task Directory
2018-09-19T19:41:19Z  Received    Task received by client
```

### Reproduction steps

Run a nomad dispatch job which is set to not restart with too little RAM allocated to it, let it get OOM killed, watch it get restarted.

### Job file (if appropriate)

```
job "SALMON_0_8192" {
  datacenters = ["dc1"]

  type = "batch"
  priority = 50

  parameterized {
    payload       = "forbidden"
    meta_required = [ "JOB_NAME", "JOB_ID"]
  }

  group "jobs" {
    restart {
      attempts = 0
      mode = "fail"
      # delay    = "30s"
    }

    ephemeral_disk {
      size = "10"
    }

    task "salmon" {
      driver = "docker"

      # This env will be passed into the container for the job.
      env {
        REDACTED
      }

      # The resources the job will require.
      resources {
        # CPU is in AWS's CPU units.
        cpu = 1024
        # Memory is in MB of RAM.
        memory = 8192
      }

      logs {
        max_files = 1
        max_file_size = 1
      }

      constraint {
        attribute = "${meta.volume_index}"
        operator  = "="
        value     = "0"
      }

      config {
        image = "wkurt/dr_salmon:OOM"
        force_pull = false

        # The args to pass to the Docker container's entrypoint.
        args = [
          "python3",
          "manage.py",
          "run_processor_job",
          "--job-name", "${NOMAD_META_JOB_NAME}",
          "--job-id", "${NOMAD_META_JOB_ID}"
        ]

        volumes = ["/var/ebs:/home/user/data_store"]

        logging {
          type = "awslogs"
          config {
            awslogs-region = "us-east-1",
            awslogs-group = "data-refinery-log-group-kurtwheeler-dev",
            awslogs-stream = "log-stream-processor-docker-kurtwheeler-dev"
          }
        }
      }
    }
  }
}
```

**Comment by dadgar** [comment]
Hey the job is being rescheduled because it failed. To disable rescheduling, set the reschedule attempts to zero: https://www.nomadproject.io/docs/job-specification/reschedule.html#attempts

**Comment by kurtwheeler** [comment]
Ok, I guess that makes sense then. I actually looked at that page though and saw this:

```
Nomad will attempt to schedule the task on another node if any of its allocation statuses become "failed". It prefers to create a replacement allocation on a node that hasn't previously been used.
```

which made me think that this wasn't what was affecting me because:
1. I hadn't specified any rescheduling parameters.
2. The task was being requeued directly on the same node and the documentation said it would be scheduled on another node.

Anyway, I'm glad there's an easy solution but I think perhaps the documentation around rescheduling could be improved because I didn't know it was a feature of Nomad until I saw the page and after I did I was pretty sure it didn't apply to me. I'll admit I didn't read the entire page, but there's a lot of subheaders under `Job Specification` and I don't think I've seen the reschedule stanza mentioned outside of its page so I assumed there weren't default values being set.

Also I'll admit that now that I know `reschedule` is a thing, I am very confused as to the difference between `reschedule` and `restart`.
